### PR TITLE
Separate raw value representations of postgresql and mysql backend

### PR DESCRIPTION
**Pull request by weiznich**
This commit separates the raw value representation of the mysql and
postgresql backend from raw `&[u8]` to specific types that could hold
additional information. To avoid code duplication for commonly used
types an additional trait (`BinaryRawValue`) is introduced, which is
basically equivalent to `AsRef<[u8]>` but has the existing
`HasRawValue` trait as super trait. This is required to workaround
some higher ranked life time issues in rustc.
Additionally some type information are added to the postgresql value
type. Those information may allow to do dynamic type conversations in
future diesel versions.

Basically a improved reimplementation of #1833 on top of #1976. As @sgrif stated there this PR keeps the common impl's in place by adding a `BinaryRawValue` trait.

That said: I'm not super sure if the newly added public types/functions have meaningful names or if we should choose something different.

cc @pgab

### CODE REVIEW

**Comment by weiznich** [comment]
All remaining build failures are fixed by #2131

**Comment by weiznich** [comment]
@sgrif Stated that he has some comments on this.

**Comment by weiznich** [comment]
@sgrif I've done the changes we have discussed some weeks ago.

**Comment by weiznich** [comment]
@JohnTitor Would it possible to get a final review on this (assuming that the CI now succeeds)?
Seans comments are now addressed.
